### PR TITLE
Feature #9 Implementation: Specify Slides Subfolder

### PR DESCRIPTION
**Pull request by etbrow**
This request is to implement feature request #9 

I modified the open files function to read the specified sub-directory of the root folder based on the slide in the settings. Doing it this way reduced the chance that a new .vscode file was created in the slides directory. The actual process should be pretty straight-forward but I can add more documentation in-code as necessary.

I've tested this on Windows only

### CODE REVIEW

**Comment by nicoespeon** [comment]
@etbrow can you rebase your branch on top of current master and fix the conflicts?

Let me know if you need help for that.

**Comment by etbrow** [comment]
I'm having some issues getting the automated tests to work so this may take a little bit.

**Comment by nicoespeon** [comment]
@etbrow yes I can help, thanks for raising the question. That's a very good pretext for me to explain the structure a bit, so everything will make more sense.

## The long story

I have separated the "domain" part from the rest—extension and something I called "adapter", I'll get back to it.

In the domain, I put the "business logic". Because the extension behaviour is not really complex, business logic tend to be a simple orchestration like the `next()` function: "if slides is active, close the markdown preview, go to the next slide and open the preview if it's a markdown file again": https://github.com/nicoespeon/vscode-slides/blob/3485a2ed5deead83bd083f14f355d9a5bd1d8208/src/domain.ts#L26-L36

The domain doesn't know about VS Code, or the way I actually store the data. It doesn't know because I pass `editor` and `repository` as parameters. These have contracts (e.g. `repository` should have a method `store()`) but it can really be anything that fulfills the contracts.

That allows me to test the business logic independently from VS Code API or the file storage. It makes tests faster and I don't have to deal with flaky things like "there can only be one instance of VS Code running yada yada yada". These are what I consider unit tests.

In the `extension.ts` file, I'll actually provide an `editor` and a `repository` that uses VS Code. The extension code is super small and simply connects things together, so I'm fine not testing it.

Finally there is the adapters. These are the concrete implementations of the contracts, using VS Code API. I try to make these as small as possible. I translate what `store()` means with the VS Code API: https://github.com/nicoespeon/vscode-slides/blob/3485a2ed5deead83bd083f14f355d9a5bd1d8208/src/adapters.ts#L150-L160

Hopefully, they don't change often once written.

Ideally, I'd like to test these. To do so, yes, I'd need to use the VS Code API in tests. So these tests will be slower and more flaky. These are what I called integration tests. But because I didn't manage to get VS Code API work with the Jest runner, I decided not to write these tests—until this problem is fixed. I detailed that in https://github.com/nicoespeon/vscode-slides/blob/master/adrs/0002-no-integration-test.md

It's not ideal that I don't have integration tests, but that's OK because most of the code logic is still tested with unit tests. That means I need to manually test when I do a change on the adapter code. The good news is that I don't have to often change the adapter code. 

## What that means for the PR

We should not import VS Code API directly in the domain. Domain doesn't know about VS Code at all. It could be any `editor`, as long as it provides the capabilities of the `Editor` contract.

We need to extract the VS Code API logic back in the adapter.

To do so, think about the method we'd like to have in theory in the `editor`. Use that in the domain (and in the tests). Then implement the method in the adapter. There should not be much logic in such method. Adapter method implementation should stay as straightforward as possible: https://github.com/nicoespeon/vscode-slides/blob/3485a2ed5deead83bd083f14f355d9a5bd1d8208/src/adapters.ts#L25-L39

I hope that will clarify things.

**Comment by nicoespeon** [comment]
However, I'm surprised that the PR diff still show changes that are already in `master`. I'll try to give it a look, see how I can get the change you made on top on the previous PR and start from there 👍

**Comment by etbrow** [comment]
Sounds good. This is a busy week for me but I'll try and get some work done on it. Makes sense, though.

File: src/domain.ts [hunk 1]
```diff
@@ -91,13 +98,60 @@ function getSlidesSettings(editor: Editor): Settings {
...
+    }),
+    ...(slidesFolder && { "workbench.slidesFolder": slidesFolder })
   });
 }
 
+/*
+ *   The previous version of this method used the editor interface's
+ *   open all files method to open the files in the root folder. To
+ *   make sure that another settings file wasn't created, I had to
+ *   use a more manual process.
```
Comment (nicoespeon) [inline src/domain.ts]: Thanks for the comment, it helps understanding the rationale behind the code.

So, in fact, the `editor` is my own adapter. It won't create another settings file or anything. It uses VS Code API to open all files, manually as you said. I think we should change `openAllFiles()` signature to take a `Folder` parameter. And update the implementation so it actually open files that are in that folder. But I think the final change will be quite small.

File: src/domain.ts [hunk 2]
```diff
@@ -120,6 +174,7 @@ interface Configuration {
   theme: string | null | undefined;
   fontFamily: string | null | undefined;
   previewMarkdownFiles: boolean;
+  slidesFolder: string | "" | undefined;
```
Comment (nicoespeon) [inline src/domain.ts]:
```suggestion
  slidesFolder: string | undefined;
```

Since it would be redundant with the `string` type ;)

**Pull request merged**